File: ruby-call-signature/src/main/java/org/jetbrains/ruby/codeInsight/types/signature/RSignatureContractContainer.java
package org.jetbrains.ruby.codeInsight.types.signature;

import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.logging.Logger;

public class RSignatureContractContainer {
    private static final Logger LOGGER = Logger.getLogger(RSignatureContractContainer.class.getName());

    @NotNull
    private final Map<MethodInfo, RSignatureContract> myContracts;
    @NotNull
    private final Map<MethodInfo, Integer> myNumberOfCalls;

    public RSignatureContractContainer() {
        myContracts = new HashMap<>();
        myNumberOfCalls = new HashMap<>();
    }

    public void reduce()
    {
        myContracts.entrySet().stream()
                .sorted(Collections.reverseOrder(Comparator.comparingInt(entry -> myNumberOfCalls.get(entry.getKey()))))
                .forEachOrdered(entry -> {
                    final MethodInfo key = entry.getKey();
                    final RSignatureContract contract = myContracts.get(key);
                    synchronized (contract) {
                        contract.minimize();
                    }
                });
        LOGGER.fine("Finished");
    }

    public void addContract(@NotNull MethodInfo info, @NotNull RSignatureContract contract) {
        myContracts.put(info, contract);
    }

    public boolean acceptTuple(@NotNull RTuple tuple) {
        MethodInfo currInfo = tuple.getMethodInfo();

        if (myContracts.containsKey(currInfo)) {
            RSignatureContract contract = myContracts.get(currInfo);

            return tuple.getArgsInfo().equals(contract.getArgsInfo()) && myContracts.get(currInfo).accept(tuple);
        } else {
            return false;
        }
    }

    public void addTuple(@NotNull RTuple tuple) {
        MethodInfo currInfo = tuple.getMethodInfo();

        if (myContracts.containsKey(currInfo)) {
            RSignatureContract contract = myContracts.get(currInfo);

            synchronized (contract) {
                if (tuple.getArgsInfo().size() == contract.getArgsInfo().size()) {
                    contract.addRTuple(tuple);
                    myNumberOfCalls.compute(currInfo, (method, oldNumber) -> oldNumber != null ? oldNumber + 1 : 1);
                }
            }

        } else {
            RSignatureContract contract = new RSignatureContract(tuple);
            myContracts.put(currInfo, contract);
        }
    }

    public Set<MethodInfo> getKeySet() {
        return myContracts.keySet();
    }

    public RSignatureContract getSignature(MethodInfo info) {
        return myContracts.get(info);
    }
}

File: ruby-call-signature/src/main/java/org/jetbrains/ruby/codeInsight/types/signature/RSignatureContractContainer.kt
package org.jetbrains.ruby.codeInsight.types.signature

class RSignatureContractContainer {

    private val myContracts: MutableMap<MethodInfo, RSignatureContract> = HashMap()
    private val myNumberOfCalls: MutableMap<MethodInfo, Int> = HashMap()

    fun acceptTuple(tuple: RTuple): Boolean {
        val currInfo = tuple.methodInfo

        val contract = myContracts[currInfo]
        return contract != null && tuple.argsInfo == contract.argsInfo && SignatureContract.accept(contract, tuple)
    }

    fun addTuple(tuple: RTuple) {
        val currInfo = tuple.methodInfo

        if (myContracts.containsKey(currInfo)) {
            val contract = myContracts[currInfo]

            if (tuple.argsInfo.size == contract?.argsInfo?.size) {
                contract.addRTuple(tuple)
                myNumberOfCalls.compute(currInfo) { _, oldNumber -> (oldNumber ?: 0) + 1 }
            }
        } else {
            val contract = RSignatureContract(tuple)
            myContracts.put(currInfo, contract)
        }
    }

    val registeredMethods: Set<MethodInfo>
        get() = myContracts.keys

    fun getSignature(info: MethodInfo): RSignatureContract? {
        return myContracts[info]?.apply { minimize() }
    }

    fun clear() {
        myContracts.clear()
    }

    val size: Int
        get() = myContracts.size
}